AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.java

```diff
@@ -1048,7 +1053,60 @@ public Uri insert(Uri uri, ContentValues values) {
             case DECKS_ID:
                 // Deck ID is generated automatically by libanki
                 throw new IllegalArgumentException("Not possible to insert deck with specific ID");
+            case MEDIA:
+                // insert a media file
+                // contentvalue should have data and preferredFileName values
+                Uri fileUri = Uri.parse(values.getAsString(FlashCardsContract.AnkiMedia.FILE_URI));
+                String preferredName = values.getAsString(FlashCardsContract.AnkiMedia.PREFERRED_NAME);
+                InputStream inputStream = null;
 
+                try {
+                    ContentResolver cR = mContext.getContentResolver();
+                    Media media = col.getMedia();
+                    // idea, open input stream and save to cache directory, then
+                    // pass this (hopefully temporary) file to the media.addFile function.
+                    inputStream = cR.openInputStream(fileUri);
+                    String fileMimeType = MimeTypeMap.getSingleton().getExtensionFromMimeType(cR.getType(fileUri)); // return eg "jpeg"
+                    // should we be enforcing strict mimetypes? which types?
+                    File tempFile;
+                    try {
+                        tempFile = File.createTempFile(
+                                preferredName+"_", // the beginning of the filename.
+                                "." + fileMimeType, // this is the extension, if null, '.tmp' is used, need to get the extension from MIME type?
+                                new File(media.dir())
+                        );
+                        tempFile.deleteOnExit();
+                    } catch (Exception e) {
+                        Timber.e(e, "Could not create temporary media file. ");
+                        return null;
+                    }
+
+                    // copy contents into temp file (possibly check file size and warn if large?)
+                    try {
+                        CompatHelper.getCompat().copyFile(inputStream, tempFile.getAbsolutePath());
+                    } catch (FileNotFoundException e) {
+                        Timber.e(e, "File not found when opening stream for supplied media file.");
+                        return null;
+                    } catch (Exception e) {
+                        Timber.e(e, "Unable to copy media file from ContentProvider");
+                        return null;
+                    } finally {
+                        assert inputStream != null;
+                        inputStream.close();
```

Comment:
assert doesn't work in Java, it is typically a no-op but if they ever fix it, then this will crash the app which is a terrible outcome, this should be

```suggestion
                        if (inputStream != null) {
                            inputStream.close();
                        }
```